test/Test/Pos/Util/ModifierSpec.hs

```diff
@@ -0,0 +1,185 @@
+                (insertDeleteThenLookup @Text @Text)
+        describe "keys" $ do
+            prop
+                "getting the keys from a map modifier does not eliminate any from the\
+                \ produced list unless they are in the argument list and are a key in the\
+                \ map modifier simultanously"
+                (allKeysAreInMap @Text @Text)
+        describe "values" $ do
+            prop
+                "returns all values from underlying storage except deleted values,\
+                \ overrides values according to insertions into MapModifier and it also\
+                \ returns new values inserted into MapModifier"
+                (allValsAreInMap @Text @Text)
+        describe "toList" $ do
+            prop
+                "getting the contents of the map modifier as an association list will\
+                \ keep all of the keys and will include the pairs in the argument list\
+                \ whose key is not present in the modifier"
+                (allPairsAreInAssocList @Text @Text)
+        describe "insertionsMap" $ do
+            prop
+                "the created hashmap removes all the deleted keys from the 'MapModifier'\
+                \ and keeps all the inserted keys"
+                (mapModifierToHashMap @Text @Text)
+        describe "insertions" $ do
+            prop
+                "the created list removes all the deleted keys from the used\
+                \ 'MapModifier' and keeps all the inserted keys in the association list"
+                (mapModifierToList @Text @Text)
+        describe "deletions" $ do
+            prop
+                "the created list of keys removes all the inserted values and keeps those\
+                \ which have been deleted"
+                (mapModifierToDeletionList @Text @Text)
+        describe "mapMaybe" $ do
+            prop
+                "the created list of keys removes all the deleted values and keeps the\
+                \ pairs from the input list whose keys are not in the modifier"
+                (allPairsAreInMaybeList @Text @Text @Text Just)
+        describe "modifyHashMap" $ do
+            prop "modifying a hashmap with a map modifier removes the keys that were\
+                 \ deleted and keeps those which were inserted, along with their values"
+                 (hashMapIsModified @Text @Text)
+
+
+mapModifierFormsMonoid
+    :: (Show k, Show v, Eq k, Eq v, Hashable k)
+    => Core.MapModifier k v
+    -> Core.MapModifier k v
+    -> Core.MapModifier k v
+    -> Property
+mapModifierFormsMonoid = formsMonoid
+
+insertThenLookup
+    :: (Show k, Show v, Eq k, Eq v, Hashable k)
+    => k
+    -> v
+    -> v
+    -> HM.HashMap k v
+    -> Core.MapModifier k v
+    -> Property
+insertThenLookup k v1 v2 hm mapMod =
+    let newMap = Core.insert k v1 mapMod
+        newHMap = HM.insert k v2 hm
+    in v1 /= v2 ==> maybe False (== v1) $ Core.lookup (flip HM.lookup newHMap) k newMap
```

Comment:
It's not, but I wrote it to emphasize the fact that the values don't have to be equal for this property to hold. In fact, them being equal is the case we don't care about because the inner `flip HM.lookup newHMap` is going to make it work anyway.